Date calculator [updating daysBetween goodie]

GreatDanton opened the pull request:
New pull request, old one had screwed commit history.

old pr: https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2734

https://duck.co/ia/view/days_between
Maintainer: @JetFault

daxtheduck commented:
@GreatDanton Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

GreatDanton commented:
oops, seems like we changed title at the same time @GuiltyDolphin

mintsoft commented on share/goodie/days_between/days_between.js:
```diff
@@ -0,0 +1,308 @@
+  console.log("DDH.days_between.build"); // remove this before submitting pull request
+
+  // functions
+  function inYears(startDate, endDate) {
```
Minor thing; can you format this to 4 spaces per tab? :+1:

mintsoft commented on t/DaysBetween.t:
```diff
@@ -12,141 +12,141 @@ zci is_cached   => 1;
-            result    => 53
-        },
-    ),
+#     'days between 01/01/2000 01/01/2001' => test_zci(
```
If we're going to move all the logic out of the Perl; I really feel we should get some solid tests around the JS implementation @moollaza Any thoughts?

GuiltyDolphin commented:
@GreatDanton There's an *awful* lot of work going on in that JS - work that would be *much* more suited to Perl (where you even get to take advantage of some of our Roles, and localization!); is there any particular reason you've moved it out into the JS rather than just done calls to the Perl? Even if you want this in the JS I'd really recommend using `moment.js` instead of rolling your own parsing and calculating; most of the logic is redundant.

GreatDanton commented:
@GuiltyDolphin 

> is there any particular reason you've moved it out into the JS rather than just done calls to the Perl?
Yes, I am not familiar with Perl.

I tried to use moment.js, just like @moollaza suggested in one of the emails, but I couldn't get it to work (link: https://gist.github.com/moollaza/1a5000c4f8e0f9fcebfc47220f479ecb). If you can figure out what is wrong in there I would be glad to use moment.js instead of my own functions.

GuiltyDolphin commented:
@GreatDanton I wouldn't worry about not being too familiar with Perl - we're all here to help with the learning experience!

As they say:

```
    'Tis a lesson you should heed:
    Try, try, try again.
    If at first you don't succeed,
    Try, try, try again.
```

I'll have a look - but regardless of whether we use Perl or moment.js, I don't think you should be rolling your own ("Don't re-invent the wheel").

GuiltyDolphin commented:
@GreatDanton What's the issue in the snippet? You don't seem to have really utilized moment in it (you still have many additionally-defined functions). Were you getting errors?

GreatDanton commented:
First the template was not rendering, so I changed` DDH.days_between.build_async` to `DDH.days_between.build`. After that template was rendering, but the calculations were not working (there were no errors in console). No matter what I tried calculations were not working, until I removed those two new lines
>DDG.require('moment.js', function () {
>		DDH_async_add({

After that everything was working like it should, so I left it like that and sent an email, but unfortunately didn't get any reply back :)

GuiltyDolphin commented:
@GreatDanton Why do you need `DDH_async_add`? `DDG.require` should be fine.

GreatDanton commented:
@GuiltyDolphin I just copied the implementation from the link I posted above. 

I removed that unnecessary part `DDH_async_add` and can see moment.js get request in console, however code inside onShow: function() does not work anymore.

GuiltyDolphin commented:
@GreatDanton Here, have a look at [DateMath](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2897/files) for an example - you want the `DDG.require` within `onShow`.

GreatDanton commented:
@GuiltyDolphin thanks, that did the trick.

moollaza commented:
@GreatDanton as @GuiltyDolphin said, we should leverage Moment.js here because it will handle nearly all the work.

When loading JS assets in a Goodie you need to use the `async` functions.

I just pulled down your PR and fiddled with the code and got it working.

I've updated my Gist with the working code: https://gist.github.com/moollaza/1a5000c4f8e0f9fcebfc47220f479ecb

Please put that into the PR and we can push this to beta.

We can definitely remove a lot of the code here by using moment.js functions.

GreatDanton commented:
@moollaza I see there is toObject() and toArray() functions which can reduce the code a lot.

GreatDanton commented:
@moollaza I tried to replace as much code as possible with moment.js functions.

Is it possible to write it in even less lines?

moollaza commented on share/goodie/days_between/days_between.js:
```diff
@@ -0,0 +1,176 @@
+                  endYear = end[0];
+                }
+
+                // insert numbers into text boxes
```
Can't we initialize the values and pass them along to the template to render, instead of filling them after display?

moollaza replied:
Looks like this isn't actually working?

I'm getting empty fields and `NaN`:

<img width="914" alt="number_of_days_between_january_15_2016_and_april_1_2016_at_duckduckgo" src="https://cloud.githubusercontent.com/assets/873785/15576785/8c4aeab6-2327-11e6-9a9c-2ce1b8c658a9.png">

GreatDanton commented:
@moollaza yeah, it was working before I deleted those functions. I will try to rewrite them in simpler way.

moollaza commented on share/goodie/days_between/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+			</form>
+			<form class="frm  frm--vrt">
+				<label class="frm__label">Month</label>
+				<input type="tel" maxlength="2" id="zci--days_between--end_month" class="frm__input" placeholder="mm" autocomplete="off" />
```
let's use `type=number` -- then we can set min/max values and don't need to validate with JS

moollaza commented on share/goodie/days_between/days_between.js:
```diff
@@ -48,23 +48,47 @@ DDH.days_between.build_async = function (ops, DDH_async_add) {
+        });
+
+        // if query does not contain numbers, return
+        if (dates.length < 1) {
```
This should be checked in the Perl -- no point in loading this JS and displaying the IA if the query is not valid 👍

moollaza commented on share/goodie/days_between/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+<div class="fix">
+	<div class="half zci--days_between--centered">
+		<div class="zci--days_between--inline">
```
Your classes are technically fine, but they're overly verbose IMO. We already wrap your IA in `<div class="zci__days_between">`. You can use that to namespace all your CSS selectors.

That way you can use classes in here to specify the content, or use generics like `class="centered"`

Alternatively I would use classes that describe the element: `class="days_between--container"`

GreatDanton replied:
I searched for .zci--days_between wrapper, but didn't find any, do I have to manually add it?

moollaza commented on share/goodie/days_between/days_between.css:
```diff
@@ -0,0 +1,26 @@
+#zci--days_between--start_day, #zci--days_between--start_month, #zci--days_between--end_day, #zci--days_between--end_month {
```
- each selector should be on a new line
- we typically avoid ID's in favour of classes

moollaza commented on share/goodie/days_between/days_between.css:
```diff
@@ -0,0 +1,26 @@
+}
+.zci--days_between--centered {
+    flex-flow: column;
+    display: flex;
```
I don't believe Flex is supported in IE 9 & 10? We need to ensure IE9+ compliance so let's stick to standard CSS

moollaza commented on share/goodie/days_between/days_between.css:
```diff
@@ -0,0 +1,26 @@
+#zci--days_between--start_year, #zci--days_between--end_year {
+    width: 55px;
+}
+.zci--days_between--inline {
```
I would drop this rule and use the grid percentage classes to ensure the `<form>` elements are aligned:

```html
<div class="zci__body">
    <h3 class="c-list__title">Date Math</h3>
    <div class="gw">
        <div class="g half ">
            <span class="h-m forty">Start Date: </span>
            <form class="frm  frm--vrt twenty">
                <label class="frm__label">Day</label>
                <input type="number" max="31" class="start_day frm__input" placeholder="dd" autocomplete="off">
            </form>
            <form class="frm  frm--vrt twenty">
                <label class="frm__label">Month</label>
                <input type="number" max="12" class="start_month frm__input" placeholder="mm" autocomplete="off">
            </form>
            <form class="frm  frm--vrt twenty">
                <label class="frm__label">Year</label>
                <input type="number" max="9999" class="start_year frm__input" placeholder="yyyy" autocomplete="off">
            </form>
        </div>

        <div class="g half">
            <span class="h-m forty">End Date: </span>
            <form class="frm  frm--vrt twenty">
                <label class="frm__label">Day</label>
                <input type="number" max="31" class="end_day frm__input" placeholder="dd" autocomplete="off">
            </form>
            <form class="frm  frm--vrt twenty">
                <label class="frm__label">Month</label>
                <input type="number" max="12" class="end_month frm__input" placeholder="mm" autocomplete="off">
            </form>
            <form class="frm  frm--vrt twenty">
                <label class="frm__label">Year</label>
                <input type="number" max="9999" class="end_year frm__input" placeholder="yyyy" autocomplete="off">
            </form>
        </div>
    </div>
```

I believe that works.. (I was fiddling with the DOM)

moollaza replied:
This should remove the need for Flex and some of the extra CSS classes

moollaza replied:
Actually I need the design team to sign off on this so for now you can keep your layout as-is, but let's remove the flexbox so it all works in IE 9.

I suspect the design team may ask to reorganize the subtitles + inputs

moollaza commented on lib/DDG/Goodie/DaysBetween.pm:
```diff
@@ -1,41 +1,32 @@
+                   options => {
+                     subtitle_content => 'DDH.days_between.content',
+                     content => "record",
+                     rowHighlight => 'true',
```
Whoops, I thought we had removed this! The design team no longer wants us using the `rowHighlight`

moollaza commented on share/goodie/days_between/days_between.js:
```diff
@@ -0,0 +1,173 @@
+
+
+    // output numbers to table
+    function writeToTable(outputArray) {
```
Seeing as we're using a record template, we should be able to use our `rerender` functionality with `item.set({ ... })` which should dynamically update the HTML so you don't need to modify the DOM to update the table

moollaza commented on share/goodie/days_between/days_between.js:
```diff
@@ -0,0 +1,173 @@
+    // return seconds, minutes, hours, days, years between starting and
+    // ending date
+    function returnValues(startingDate, endingDate) {
+        var seconds = Math.abs(startingDate.diff(endingDate, 'seconds'));
```
nit: we could avoid the `Math.abs` by verifying `startingDate.isBefore(endingDate)` and if not flip the values

moollaza commented on share/goodie/days_between/days_between.js:
```diff
@@ -0,0 +1,173 @@
+
+    // return seconds, minutes, hours, days, years between starting and
+    // ending date
+    function returnValues(startingDate, endingDate) {
```
I would use an object where the keys are seconds, minutes, hours..

Then loop over that and set the value for the key to the diff, specifying the key name as the param:

```js
var out = { 
 seconds: 0,
 minutes: 0,
 ...
};

$.foreach(out, function(k,v) {
 out[k] = Math.abs(startingDate.diff(endingDate, v);
});
```

Much more succint :)

GreatDanton replied:
> var out = { 
>  seconds: 0,
>  minutes: 0,
>  ...
> };
> 
> $.foreach(out, function(k,v) {
>  out[k] = Math.abs(startingDate.diff(endingDate, v);
> });

This is brilliant ^^. Added, will use it once I figure out rerender stuff.

moollaza commented on lib/DDG/Goodie/DaysBetween.pm:
```diff
@@ -1,41 +1,32 @@
+               data => {
+                 title => "Date Math",
+                   },
+               templates => {
```
Seeing as you specify the templates obj in the JS you can simply set `template => 1` or `templates => {}` in the Perl

moollaza commented on share/goodie/days_between/days_between.js:
```diff
@@ -0,0 +1,173 @@
+                $('#zci--days_between--calculate').click(function () {
+
+                    // get numbers from text boxes
+                    startDay = $('#zci--days_between--start_day').val();
```
These selectors should be cached so each time the button is clicked we're not re-parsing the DOM. 

I would do something like:

```js
var $daysBetween = $(.zci--days_between),
$calculateBtn = $daysBetween.find(".calculate"),
$startDay = $daysBetween.find(".start_day"),
$startMonth = $daysBetween.find(".start_month"),
...;

$calculatBtn.click(function() {
 startDay = $startDay.val();
 ...
});
```

If you use an object to hold onto your selectors and values you can probably use a loop to grab all the vals so the code is less verbose

GreatDanton commented:
![ddg_new](https://cloud.githubusercontent.com/assets/11252004/15610456/9f37edc8-2425-11e6-89ef-5354ade30037.png)

We can now use month names in query, however if I try to use 

> days between 10.1.2000 and 10 feb 2000
or 
> days between january 10 2000 and february 10 2000

it doesn't work anymore. Guess I will have to fiddle some more with the code. At least it's a bit more clear what the code is doing right now.

moollaza commented:
@GreatDanton I demoed this IA for the team today and they all liked it, however they brought up a great point -- it's not clear that the different measurements of time are independent of each other.

Meaning Is the difference between the two dates in that image exactly 2 years? Or is it two years and X days and X hours (if times are given).

Their initial thought was that it was a total of 2 years + 864 days + 20737 hours ...

I'll have them review the design and offer some suggestions. Basically I think we need to make the absolute time difference most apparent and then indicate what the equivalent measurements of that are

E.G.

Difference = 2 years, 65 days, 18 hours 35 minutes 

Equivalent to...
2.XXX years
YYY    days
ZZZZZ hours
AAAA  minutes
...

edgesince84 commented:
@GreatDanton I wanted to see if you still planned on finishing this up - we really appreciate your contribution and want to see this go live. We're closing up old PRs in the next week, so let me know if I should leave this open for you!

If you'd rather work on something new, we'd like to invite you to our [new forum](https://forum.duckduckhack.com/) to preview how we're improving programming-related searches. There are a lot of important tasks up for grabs that we'd like your help with so feel free to and [jump in]](https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53)

GreatDanton commented:
@edgesince84,
thanks for reminder. I've been a bit tight on time in the past two months, however I would like to finish what I started.

Since this branch has conflicts now, how do I resolve this? Also apart from the additional code that I mentioned above, are there any other suggestions to push this IA to beta faster?

edgesince84 commented:
Hey @GreatDanton  - I'm going to defer those questions to my more-technical colleagues @moollaza & @tagawa

moollaza commented:
> Since this branch has conflicts now, how do I resolve this?

@GreatDanton you need to `git pull upstream master`. Then look at DaysBetween.pm and resole the merge conflict (remove anything that shouldn't be there) and commit this change. Then push it  to your branch used for this PR. After that the code should be mergeable again. Once that is done we can deploy this to beta.

@GreatDanton I think to keep this PR the IA design simple, we should consider removing the table for now, because it's unclear what the other units of time represent, and could be considered irrelevant seeing as the query specifies, in this case "days".

That way this PR will introduce the new interactivity and design.

Subsequent PRs can add additional features like adjusting the result based on the request unit of time in the query. E.g.  "seconds between X and Y" should give the number of seconds

Maybe then for the query "time between X and Y" we can show a table?

-----

As well, we now have Pikaday.js available to Instant Answers which mean we can further improve this IA with a dat selection tool meaning we don't need multiple inputs for dates and done need to bother with checking the dates for validity!

moollaza commented on t/DaysBetween.t:
```diff
@@ -10,6 +10,145 @@ use Test::MockTime qw( :all );
 zci answer_type => 'days_between';
 zci is_cached   => 1;
 
+<<<<<<< HEAD
```
@GreatDanton looks like part of the merge conflict still remains. You need to resolve the conflicts in this files as well, and then it should be good 👍

moollaza commented:
@GreatDanton thanks for the updates! Deploying to Beta for testing...

moollaza commented on t/DaysBetween.t:
```diff
@@ -10,44 +10,8 @@ use Test::MockTime qw( :all );
-
 ddg_goodie_test(
     [qw( DDG::Goodie::DaysBetween)],
-    'days between today and tomorrow'                            => build_test('14 Jul 2016', '15 Jul 2016', 1, ''),
```
@GreatDanton we should at least test that all these queries trigger the IA and the ones that shouldn't return a result still return `undef`.

It looks like this test file is causing problems with Beta so I'm unable to get this live on Beta

mintsoft commented:
@GreatDanton Hiya, just checking in. Are you OK to finish this off? It's pretty close to being ready now. If not we'll open it up to the community to finish

GreatDanton commented:
@mintsoft ah, yes winter school semester took my entire free time. Thanks for reminder. I will try to finish this, since I have a bit more free time now.

GreatDanton commented:
According to the [docs](https://docs.duckduckhack.com/testing-reference/test-files.html), I should provide plain text response + html response. I can copy and paste html response in test, but I am not sure how plain text response should look like.

On the other side, I saw some js heavy goodies having similar test I provided. So what is the way to go for writing tests in my case?

mintsoft commented:
@GreatDanton Plaintext responses are usually a statement of input and output in one go like: "19-Jan-2017 + 3 days =  22-Jan-2017"

moollaza commented:
This PR is being closed due to inactivity. If you're still interested in finishing, please let me know and I'll gladly re-open it for you!
